File: jicofo-common/src/main/java/org/jitsi/impl/protocol/xmpp/ChatRoom.java
/*
 * Jitsi, the OpenSource Java VoIP and Instant Messaging client.
 *
 * Copyright @ 2015-Present 8x8, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jitsi.impl.protocol.xmpp;

import org.jetbrains.annotations.*;
import org.jitsi.jicofo.xmpp.muc.*;
import org.jitsi.utils.*;
import org.jivesoftware.smack.*;
import org.jivesoftware.smack.packet.*;
import org.jxmpp.jid.*;

import java.util.*;

/**
 * Represents a chat channel/room/rendez-vous point/ where multiple chat users
 * could rally and communicate in a many-to-many fashion.
 *
 * @author Emil Ivov
 * @author Yana Stamcheva
 * @author Valentin Martinet
 */
public interface ChatRoom
{
    /**
     * Joins this chat room with the nickname of the local user so that the
     * user would start receiving events and messages for it.
     */
    void join()
        throws SmackException, XMPPException, InterruptedException;

    @NotNull XmppProvider getXmppProvider();

    /**
     * Returns true if the local user is currently in the multi user chat
     * (after calling one of the {@link #join()} methods).
     *
     * @return true if currently we're currently in this chat room and false
     * otherwise.
     */
    boolean isJoined();

    /**
     * Leave this chat room. Once this method is called, the user won't be
     * listed as a member of the chat room any more and no further chat events
     * will be delivered.
     */
    void leave();

    /**
     * Returns the local user's role in the context of this chat room or
     * <tt>null</tt> if not currently joined.
     *
     * @return the role currently being used by the local user in the context of
     * the chat room.
     */
    MemberRole getUserRole();

    /**
     * Returns a <tt>List</tt> of <tt>ChatRoomMember</tt>s corresponding to all
     * members currently participating in this room.
     *
     * @return a <tt>List</tt> of <tt>ChatRoomMember</tt> instances
     * corresponding to all room members.
     */
    List<ChatRoomMember> getMembers();

    /**
     * Returns the number of members that are currently in this chat room. Note that this does not include the MUC
     * occupant of the local participant.
     */
    int getMembersCount();

    /**
     * Returns the number of members that currently have their audio sources unmuted.
     */
    int getAudioSendersCount();

    /**
     * Returns the number of members that currently have their video sources unmuted.
     */
    int getVideoSendersCount();

    /**
    * Grants ownership privileges to another user. Room owners may grant
    * ownership privileges. Some room implementations will not allow to grant
    * ownership privileges to other users. An owner is allowed to change
    * defining room features as well as perform all administrative functions.
    *
    * @param member the member to grant ownership to.
    */
    void grantOwnership(@NotNull ChatRoomMember member);

    /**
     * Gets the name of this chat room as a JID.
     * @return the name of this chat room as a JID.
     */
    EntityBareJid getRoomJid();

    /**
     * Finds chat member for given MUC jid.
     *
     * @param mucJid full MUC jid of the user for whom we want to find chat
     *               member instance. Ex. chatroom1@muc.server.com/nick1234
     *
     * @return an instance of <tt>XmppChatMember</tt> for given MUC jid or
     *         <tt>null</tt> if not found.
     */
    ChatRoomMember getChatMember(EntityFullJid mucJid);

    /**
     * @return the list of all our presence {@link ExtensionElement}s.
     */
    Collection<ExtensionElement> getPresenceExtensions();

    /**
     * Checks if a packet extension is already in the presence.
     *
     * @param elementName the name of XML element of the presence extension.
     * @param namespace the namespace of XML element of the presence extension.
     *
     * @return <tt>boolean</tt>
     */
    boolean containsPresenceExtension(String elementName, String namespace);

    /**
     * Modifies our current MUC presence by adding and/or removing specified
     * extensions. The extension are compared by instance equality.
     * @param toRemove the list of extensions to be removed.
     * @param toAdd the list of extension to be added.
     */
    void modifyPresence(Collection<ExtensionElement> toRemove, Collection<ExtensionElement> toAdd);

    /**
     * Add a [ChatRoomListener] to the list of listeners to be notified of events from this [ChatRoom].
     */
    void addListener(@NotNull ChatRoomListener listener);

    /**
     * Removes a [ChatRoomListener] from the list of listeners to be notified of events from this [ChatRoom].
     */
    void removeListener(@NotNull ChatRoomListener listener);

    void setPresenceExtension(ExtensionElement extension, boolean remove);

    /**
     * Get the unique meeting ID associated by this room (set by the MUC service).
     */
    String getMeetingId();

    /**
     * Whether A/V moderation is enabled.
     * @param mediaType the media type.
     * @return whether A/V moderation is enabled.
     */
    boolean isAvModerationEnabled(MediaType mediaType);

    /**
     * Sets new value for A/V moderation.
     * @param mediaType the media type.
     * @param value the new value.
     */
    void setAvModerationEnabled(MediaType mediaType, boolean value);

    /**
     * Updates the list of members that are allowed to unmute audio or video.
     * @param whitelists a map with string keys (MediaType.AUDIO or MediaType.VIDEO).
     */
    void updateAvModerationWhitelists(Map<String, List<String>> whitelists);

    /**
     * Checks the whitelists whether the supplied jid is allowed from a moderator to unmute.
     * @param jid the jid to check.
     * @param mediaType type of media for which we are checking.
     * @return <tt>true</tt> if the member is allowed to unmute, false otherwise.
     */
    boolean isMemberAllowedToUnmute(Jid jid, MediaType mediaType);

    /**
     * Checks whether this is a breakout room or not.
     * @return <tt>true</tt> if it is, <tt>false</tt> otherwise.
     */
    boolean isBreakoutRoom();

    /**
     * Gets the main room name (JID) when in a breakout room.
     * @return The main room JID as a string.
     */
    String getMainRoom();

    @NotNull
    OrderedJsonObject getDebugState();
}

File: jicofo-common/src/main/kotlin/org/jitsi/jicofo/xmpp/muc/ChatRoom.kt
/*
 * Jicofo, the Jitsi Conference Focus.
 *
 * Copyright @ 2015-Present 8x8, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jitsi.jicofo.xmpp.muc

import org.jitsi.jicofo.MediaType
import org.jitsi.jicofo.xmpp.RoomMetadata
import org.jitsi.jicofo.xmpp.XmppProvider
import org.jitsi.utils.OrderedJsonObject
import org.jivesoftware.smack.SmackException
import org.jivesoftware.smack.XMPPException
import org.jivesoftware.smack.packet.ExtensionElement
import org.jxmpp.jid.EntityBareJid
import org.jxmpp.jid.EntityFullJid
import org.jxmpp.jid.Jid

/** Wraps a Smack [MultiUserChat] for the purposes of jicofo. */
interface ChatRoom {
    val xmppProvider: XmppProvider

    /** The JID of the chat room. */
    val roomJid: EntityBareJid

    /** Whether the local user is currently in the multi user chat (after calling one of the [.join] methods). */
    val isJoined: Boolean

    /**
     * Returns the list of members of this [ChatRoom]. Note that this does not include a representation of the local
     * user as an occupant of the MUC.
     */
    val members: List<ChatRoomMember>

    /** The size of [members], exposed separately for performance (avoid creating a new list just to get the count) */
    val memberCount: Int

    /**
     *  The number of [members] with role VISITOR. Exposed separately for performance (avoid creating a new list
     *  just to get the count).  Also includes visitors who have joined within the last vnode-join-latency-interval,
     *  as reported by [visitorInvited]. */
    val visitorCount: Int

    /** Whether a lobby is enabled for the room. Read from the MUC config form. */
    val lobbyEnabled: Boolean

    /** Whether the visitors feature is enabled for the room. Read from the MUC config form. */
    val visitorsEnabled: Boolean?

    /** Whether the visitorsLive flag is enabled for the room. Read from the MUC config form. */
    val visitorsLive: Boolean

    /** The number of participants in the room after which new endpoints should be redirected to visitors.
     *  Read from the MUC config form. */
    val participantsSoftLimit: Int?

    val debugState: OrderedJsonObject

    /** Returns the number of members that currently have their audio sources unmuted. */
    var audioSendersCount: Int

    /** Returns the number of members that currently have their video sources unmuted. */
    var videoSendersCount: Int

    /**
     * Whether a user with a certain ID and a certain group ID would be allowed to join the main room. Note that this
     * does not actually perform authentication (i.e. if the user has a valid claim for the user ID and group ID),
     * just checks the MUC configuration.
     */
    fun isAllowedInMainRoom(userId: String?, groupId: String?): Boolean

    /**
     * Whether a user with a certain ID and a certain group ID is preferred in the main room, i.e. is explicitly listed
     * in the list of participants or moderators for the main room.
     */
    fun isPreferredInMainRoom(userId: String?, groupId: String?): Boolean

    /**
     * Joins this chat room with the preconfigured nickname. Returns the fields read from the MUC config form after
     * joining.
     */
    @Throws(SmackException::class, XMPPException::class, InterruptedException::class)
    fun join(): ChatRoomInfo

    /** Leave the chat room. */
    fun leave()

    /**
     * Grants ownership privileges to another user. Room owners may grant
     * ownership privileges. Some room implementations will not allow to grant
     * ownership privileges to other users. An owner is allowed to change
     * defining room features as well as perform all administrative functions.
     *
     * @param member the member to grant ownership to.
     */
    fun grantOwnership(member: ChatRoomMember)

    /**
     * Get the [ChatRoomMember] for a given MUC occupant jid.
     *
     * @param occupantJid full MUC jid of the user for whom we want to find chat
     * member instance. Ex. chatroom1@muc.server.com/nick1234
     */
    fun getChatMember(occupantJid: EntityFullJid): ChatRoomMember?

    /** Add all of [extensions] to our presence. */
    fun addPresenceExtensions(extensions: Collection<ExtensionElement>)

    /** Add [extension] to our presence, no-op if we already have an extension with the same QName. */
    fun addPresenceExtensionIfMissing(extension: ExtensionElement)

    /** Remove presence extensions matching the predicate [pred]. */
    fun removePresenceExtensions(pred: (ExtensionElement) -> Boolean)

    /**
     * Add [extension] to presence and remove any previous extensions with the same QName.
     *
     * Note that this always sends a presence stanza. If the intention is to only send a packet if the extension was
     * modified use [addPresenceExtensionIfMissing].
     */
    fun setPresenceExtension(extension: ExtensionElement)

    /** Add a [ChatRoomListener] to the list of listeners to be notified of events from this [ChatRoom]. */
    fun addListener(listener: ChatRoomListener)

    /** Removes a [ChatRoomListener] from the list of listeners to be notified of events from this [ChatRoom]. */
    fun removeListener(listener: ChatRoomListener)

    /** Removes all [ChatRoomListener]s. */
    fun removeAllListeners()

    /** Whether A/V moderation is enabled for [mediaType] (audio or video). */
    fun isAvModerationEnabled(mediaType: MediaType): Boolean

    /** Enable or disable a/v moderation for [mediaType]. */
    fun setAvModerationEnabled(mediaType: MediaType, value: Boolean)

    /** Updates the list of members that are allowed to unmute audio or video. */
    fun setAvModerationWhitelist(mediaType: MediaType, whitelist: List<String>)

    /** Update the value in the room_metadata structure */
    fun setRoomMetadata(roomMetadata: RoomMetadata)

    /** whether the current A/V moderation setting allow the member [jid] to unmute (for a specific [mediaType]). */
    fun isMemberAllowedToUnmute(jid: Jid, mediaType: MediaType): Boolean

    /** Re-load the MUC configuration form, updating local state if relevant fields have changed. */
    fun reloadConfiguration()

    /** Notify the chatroom that a visitor has been redirected to this room.
     */
    fun visitorInvited()

    /** Queue a task to be executed sequentially with respect to other tasks queued through this method. */
    fun queueXmppTask(runnable: () -> Unit)
}

/** Holds fields read from the MUC config form at join time, which never change. */
data class ChatRoomInfo(
    /** The meeting ID, or null if none is set. */
    val meetingId: String?,
    /** The JID of the main room if this is a breakout room, otherwise null. */
    val mainRoomJid: EntityBareJid?
)

/** Whether a user with a certain JID is allowed to unmute with any of [mediaTypes]. */
fun ChatRoom.isMemberAllowedToUnmute(jid: Jid, mediaTypes: Set<MediaType>): Boolean =
    mediaTypes.any { isMemberAllowedToUnmute(jid, it) }
